crates/torii/core/src/sql.rs

```diff
@@ -1175,9 +1198,383 @@ impl Sql {
+                vec![
+                    Argument::FieldElement(contract_address),
+                    Argument::FieldElement(from),
+                    Argument::FieldElement(to),
+                    Argument::String(u256_to_sql_string(&amount)),
+                    Argument::String(token_id.clone()),
+                    Argument::String(utc_dt_string_from_timestamp(block_timestamp)),
+                ],
+            );
+        }
+
+        // Update balances in erc20_balance table
+        {
+            // NOTE: formatting here should match the format we use for Argument type in QueryQueue
+            // TODO: abstract this so they cannot mismatch
+
+            // Since balance are stored as TEXT in db, we cannot directly use INSERT OR UPDATE
+            // statements.
+            // Fetch balances for both `from` and `to` addresses, update them and write back to db
+            let query = sqlx::query_as::<_, (String, String)>(
+                "SELECT account_address, balance FROM balances WHERE contract_address = ? AND \
+                 account_address IN (?, ?)",
+            )
+            .bind(format!("{:#x}", contract_address))
+            .bind(format!("{:#x}", from))
+            .bind(format!("{:#x}", to));
+
+            // (address, balance)
+            let balances: Vec<(String, String)> = query.fetch_all(&self.pool).await?;
+            // (address, balance) is primary key in DB, and we are fetching for 2 addresses so there
+            // should be at most 2 rows returned
+            assert!(balances.len() <= 2);
+
+            let from_balance = balances
+                .iter()
+                .find(|(address, _)| address == &format!("{:#x}", from))
+                .map(|(_, balance)| balance.clone())
+                .unwrap_or_else(|| format!("{:#63x}", crypto_bigint::U256::ZERO));
+
+            let to_balance = balances
+                .iter()
+                .find(|(address, _)| address == &format!("{:#x}", to))
+                .map(|(_, balance)| balance.clone())
+                .unwrap_or_else(|| format!("{:#64x}", crypto_bigint::U256::ZERO));
+
+            let from_balance = sql_string_to_u256(&from_balance);
+            let to_balance = sql_string_to_u256(&to_balance);
+
+            let new_from_balance =
+                if from != Felt::ZERO { from_balance.sub(amount) } else { from_balance };
+            let new_to_balance = if to != Felt::ZERO { to_balance.add(amount) } else { to_balance };
+
+            let update_query = "
+            INSERT INTO balances (id, balance, account_address, contract_address, token_id)
+            VALUES (?, ?, ?, ?, ?)
+            ON CONFLICT (id) 
+            DO UPDATE SET balance = excluded.balance";
+
+            if from != Felt::ZERO {
+                self.query_queue.enqueue(
+                    update_query,
+                    vec![
+                        Argument::String(format!("{:#x}:{:#x}", from, contract_address)),
```

Comment:
As discussed with you @lambda-0x and also with @Larkooo earlier, we should have more consistency on formatting the `Felt` with `{:#066x}` or `0x{:064x}`.

We may have in Dojo a trait we implement for `Felt` to have the whole stack formatting the same way?

Reply:
i did a sanity check on the Felt type from `starknet-core` and looks like it differs from std implementation of printing hex values, so doing `{:#064x}` does actually print `0x{64 characters padded with 0}`. so we might not need to change the implementation unless i misunderstood your point.

<img width="671" alt="image" src="https://github.com/user-attachments/assets/14ec2886-87dc-4200-8a72-49363d09cf50">